### ISSUE TITLE
Numerical correctness issue

### ISSUE THREAD

**Issue by jacogreyling** [tf-coreml/tf-coreml#131]
Hi,

I'm trying to convert a pre-trained model (TensorFlow) created by David Sandberg as part of the FaceNet (https://github.com/davidsandberg/facenet) project but I'm getting some anomalies. The project does not output classifications BUT it returns the embeddings for a particular image. I'm able to convert the model to CoreML but when I run the 'ssd_example.ipynb' playbook (obviously modified to compare the 128 measurements, instead of the box_encodings and scores) I get different values.

For instance the **tf_output_values** values are:

```
[-0.07655399 -0.12588206  0.08303334  0.02634034  0.03198989  0.02764522
  0.00205382  0.02007296 -0.18876916  0.06068632  0.01070828 -0.00832692
  0.04023512  0.03970877  0.13458191 -0.14512599 -0.02042589  0.10935931
  0.17361009 -0.0789301   0.02592264  0.1351315  -0.1275854  -0.22415063
  0.14558269 -0.06989373  0.01942037 -0.0265556  -0.06761812  0.01771478
 -0.00192919  0.07230102 -0.05849576  0.03201381  0.07697071  0.07296678
  0.1357609  -0.0764974  -0.13558555 -0.04585044 -0.03343786  0.08027246
  0.03624294 -0.07711142 -0.00611687 -0.06432545 -0.12506823  0.05517385
 -0.01395683  0.0164263   0.08837812  0.07106275 -0.02754417  0.03908188
 -0.07795331  0.12266902  0.02619525 -0.06340962  0.02195978 -0.09625895
  0.13594793 -0.031368   -0.01047055 -0.00789936  0.22155324 -0.01903406
  0.02694266 -0.20015974  0.04793613  0.06622206 -0.08667485 -0.00680321
  0.08119168 -0.04803912  0.03554194  0.04418208  0.11988591 -0.01306157
 -0.03680412  0.01063573  0.01186399 -0.08117517  0.11864755 -0.08667143
 -0.12739453 -0.05611118  0.08162383 -0.17385514 -0.03586543  0.08852623
  0.0449738  -0.16948055 -0.10812906  0.00231437  0.15089905  0.02049436
 -0.02118966  0.01898761  0.0290181  -0.10996915 -0.00501223  0.0795771
  0.01498052  0.08446263 -0.09668639  0.08318477  0.12731704  0.00297362
  0.06353725 -0.06213496 -0.00163572  0.1091258   0.09730852 -0.0848889
  0.10799909  0.12218309  0.11585088  0.06378867  0.15557306 -0.02333291
  0.01459775  0.03316046  0.03348272 -0.08761434  0.06378135  0.26579535
 -0.05884812  0.02619427]
```

whilst the **coreml_outputs** values are:

```
[-0.07655233 -0.12588233  0.08302292  0.0263503   0.03198881  0.02763469
  0.00205001  0.02007863 -0.18877722  0.06068362  0.01070389 -0.00832808
  0.0402457   0.03970525  0.13457417 -0.14513074 -0.02041985  0.10935652
  0.17361169 -0.07893046  0.02592378  0.13513304 -0.12758416 -0.2241524
  0.14559138 -0.06990062  0.01941448 -0.02653882 -0.06762731  0.0177152
 -0.00192607  0.07229953 -0.05849122  0.03201883  0.07696699  0.07297074
  0.13576084 -0.07648891 -0.13559765 -0.0458595  -0.03343242  0.08026537
  0.03624137 -0.07710776 -0.00611459 -0.0643262  -0.12506543  0.05516914
 -0.013971    0.0164208   0.0883874   0.07107809 -0.02754353  0.03908283
 -0.07795771  0.12266476  0.02619473 -0.06341161  0.02196092 -0.09626702
  0.13593471 -0.03136974 -0.01046852 -0.00789117  0.22154383 -0.01904163
  0.02694696 -0.20016906  0.04794073  0.06622028 -0.08667358 -0.00680696
  0.08119727 -0.04804014  0.03553414  0.04418483  0.11989151 -0.01305712
 -0.03680269  0.01064588  0.0118598  -0.08117118  0.11862908 -0.08666934
 -0.12739702 -0.0561312   0.08162071 -0.17385228 -0.03586386  0.08852777
  0.04497356 -0.1694769  -0.10812718  0.0023156   0.15089768  0.02049297
 -0.02118986  0.01899382  0.02902044 -0.1099713  -0.00501828  0.07957549
  0.01497809  0.08447279 -0.09669739  0.08317172  0.12731971  0.00296036
  0.06354561 -0.06212025 -0.001624    0.10912062  0.09731426 -0.0848885
  0.10800248  0.12218159  0.11585984  0.0637624   0.15556058 -0.02333151
  0.01458995  0.03315885  0.03347839 -0.0876134   0.06378786  0.26579848
 -0.05884664  0.02618424]
```

The maximum relative error on embedding is: **0.000026**.

Can someone please tell me if I'm doing something wrong??

Thx

**Comment by aseemw** [comment]
The CoreML output and TF output will never match exactly. Reasons for that:
- the default execution path for CoreML (GPU) uses lower precision (float 16)
- even when the forceCPU flag is used to use the float 32 bit path, there might be differences because of different optimized algorithms used to evaluate convolutions

**Comment by aseemw** [comment]
Yes it looks correct and you should be able to use it in your iOS app.
Although it wouldn’t hurt to first try it in python with the actual data that you expect in your app. For example if you are doing detection you can plot the bounding boxes and check whether they match reasonably with the ones you get with TF.

How much accuracy you really need depends on your application. For instance, in classification you just want the top few classes to match.
You can always use the forceCPU flag to get better accuracy.

**Comment by jacogreyling** [comment]
Thank you @aseemw, you're a star! I'll go ahead and close the issue.

**Comment by ramiro-feria-puron** [comment]
@jacogreyling Tried for a while to convert to CoreML and gave up; trying it again. I am doing this:

tfcoreml.convert(tf_model_path = '20180402-114759.pb',
                     mlmodel_path = 'Test.mlmodel',
                     output_feature_names = ['softmax:0'],
                     input_name_shape_dict = {'input:0' : [1, 160, 160, 3]})

And this is what I get:

> _Please provide the shape for the input phase_train:0 through the argument 'input_name_shape_dict'_

Could you please provide any insight on what's wrong? What's the name/shape of the input? What about the name/shape of the embedding?What do you mean by

> obviously modified to compare the 128 measurements, instead of the box_encodings and scores

Any help would be very appreciated!

**Comment by ramiro-feria-puron** [comment]
@jacogreyling ^ Any help would be very appreciated!
